sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/authentication/AccessTokenCache.java

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.containers.containerregistry.authentication;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+import reactor.core.publisher.Sinks;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+class AccessTokenCache {
```

Comment:
Ideally this class can be exposed via the experimental package and we can consume it.
@g2vinay - I will sync up with you on this.

Reply:
As per the offline discussion:
Given that ACR has a separate sts service and issues its own independent tokens with custom expiry.
It makes sense to have independent caching logic for it, as it has the potential to evolve in the future if we choose to cache in the ACR tokens per scope. (That'd be a great optimization from a performance perspective if we pull it off). 

The AAD token caching logic in azure-core can evolve too in the future and will give priority to AAD tokens specifically. So, the two implementations are not directly tied together, and having independent caches, allows to optimize and augment the caching logic best for ACR and AAD respectively.

Seems both the caches are implementation details, once ACR caching logic has stabilized and evolved we can always look into merging them together without breaking public API, if possible given their mature state at that time.

Reply:
I created an issue to track it.

---
In reply to: [597131284](https://github.com/Azure/azure-sdk-for-java/pull/19926#discussion_r597131284) [](ancestors = 597131284)